cloud/aws/actuators/machine/actuator.go

```diff
@@ -104,6 +104,73 @@ func (a *Actuator) removeStoppedMachine(machine *clusterv1.Machine, client awscl
 	return TerminateInstances(client, instances, mLog)
 }
 
+func buildEc2Filters(inputFilters []providerconfigv1.Filter) []*ec2.Filter{
+	filters := make([]*ec2.Filter, len(inputFilters))
+	for i, f := range inputFilters {
+		values := make([]*string, len(f.Values))
+		for j, v := range f.Values {
+			values[j] = aws.String(v)
+		}
+		filters[i] = &ec2.Filter{
+			Name:   aws.String(f.Name),
+			Values: values,
+		}
+	}
+	return filters
+}
+
+func getSecurityGroupsIDs(securityGroups []providerconfigv1.AWSResourceReference, client awsclient.Client, mLog log.FieldLogger) ([]*string, error) {
+	var securityGroupIDs []*string
+	for _, g := range securityGroups {
+		// ID has priority
+		if g.ID != nil {
+			securityGroupIDs = append(securityGroupIDs, g.ID)
+		} else if g.Filters != nil {
+			mLog.Debug("Describing security groups based on filters")
+			ec2Filters := buildEc2Filters(g.Filters)
+			// Get groups based on filters
+			describeSecurityGroupsRequest := ec2.DescribeSecurityGroupsInput{
+				Filters: ec2Filters,
+			}
+			describeSecurityGroupsResult, err := client.DescribeSecurityGroups(&describeSecurityGroupsRequest)
+			if err != nil {
+				mLog.Errorf("error describing security groups: %v", err)
+				return nil, fmt.Errorf("error describing security groups: %v", err)
+			}
+			for _, g := range describeSecurityGroupsResult.SecurityGroups {
+				groupID := *g.GroupId
+				securityGroupIDs = append(securityGroupIDs, &groupID)
+			}
+		}
+	}
+	return securityGroupIDs, nil
+}
+
+func getSubnetID(subnet providerconfigv1.AWSResourceReference, client awsclient.Client, mLog log.FieldLogger) (*string, error) {
+	var subnetIDs []*string
+	// ID has priority
+	if subnet.ID != nil {
+		subnetIDs = append(subnetIDs, subnet.ID)
+	} else {
+		mLog.Debug("Describing subnets based on filters")
+		ec2Filters := buildEc2Filters(subnet.Filters)
+		describeSubnetRequest := ec2.DescribeSubnetsInput{
+			Filters: ec2Filters,
```

Comment:
`ec2Filters` is not used anywhere else so it's fine to `Filters: buildEc2Filters(subnet.Filters)`

Reply:
`buildEc2Filters` looks good, can you use it in the AMI selection part as well? Lines 143-157.